compiler/src/org.graalvm.compiler.asm.aarch64/src/org/graalvm/compiler/asm/aarch64/AArch64Assembler.java

```diff
@@ -3330,44 +3277,2087 @@ public void patch(int codePos, int relative, byte[] code) {
+            FCVTZS(0b11011 << 12),
+            SCVTF(0b11101 << 12),
+            CNT(0b00101 << 12),
+
+            /* Advanced SIMD shift by immediate (C4-371). */
+            SSHLL(0b10100 << 11),
+            USHLL(0b1 << 29 | 0b10100 << 11),
+            SHL(0b01010 << 11),
+            SSHR(0b00100 << 11),
+            USHR(0b1 << 29 | 0b00100 << 11),
+
+            /* Advanced SIMD extract. (C4-356). */
+            EXT(0b00 << 22);
+
+            public final int encoding;
+
+            ASIMDInstruction(int encoding) {
+                this.encoding = encoding;
+            }
+
+        }
+
+        private final AArch64Assembler asm;
+
+        protected ASIMDAssembler(AArch64Assembler asm) {
+            this.asm = asm;
+        }
+
+        protected void emitInt(int x) {
+            asm.emitInt(x);
+        }
+
+        /**
+         * Returns whether the operation is utilizing multiple vector lanes. The only scenario when
+         * this isn't true is when performing an operation using a 64-bit register and an element of
+         * size 64 bits.
+         */
+        private static boolean usesMultipleLanes(ASIMDSize size, ElementSize eSize) {
+            return !(size == ASIMDSize.HalfReg && eSize == ElementSize.DoubleWord);
+        }
+
+        /* Helper values/methods for encoding instructions */
+
+        private static final int ASIMDImm5Offset = 16;
+        private static final int ASIMDQBitOffset = 30;
+        private static final int ASIMDSizeOffset = 22;
+
+        private final int elemSize00 = 0b00 << ASIMDSizeOffset;
+        private final int elemSize01 = 0b01 << ASIMDSizeOffset;
+        private final int elemSize10 = 0b10 << ASIMDSizeOffset;
+        private final int elemSize11 = 0b11 << ASIMDSizeOffset;
+
+        private static int elemSizeXX(ElementSize eSize) {
+            return eSize.encoding << ASIMDSizeOffset;
+        }
+
+        private static int elemSize1X(ElementSize eSize) {
+            assert eSize == ElementSize.Word || eSize == ElementSize.DoubleWord;
+            return (0b10 | (eSize == ElementSize.DoubleWord ? 1 : 0)) << ASIMDSizeOffset;
+        }
+
+        private static int elemSize0X(ElementSize eSize) {
+            assert eSize == ElementSize.Word || eSize == ElementSize.DoubleWord;
+            return (eSize == ElementSize.DoubleWord ? 1 : 0) << ASIMDSizeOffset;
+        }
+
+        /**
+         * Sets the Q-bit if requested.
+         */
+        private static int qBit(boolean isSet) {
+            return (isSet ? 1 : 0) << ASIMDQBitOffset;
+        }
+
+        /**
+         * Sets the Q-bit if using all 128-bits.
+         */
+        private static int qBit(ASIMDSize size) {
+            return (size == ASIMDSize.FullReg ? 1 : 0) << ASIMDQBitOffset;
+        }
+
+        private void asimdAcrossLanes(ASIMDInstruction instr, ASIMDSize size, int eSizeEncoding, Register dst, Register src) {
+            int baseEncoding = 0b0_0_0_01110_00_11000_00000_10_00000_00000;
```

Comment:
I did it because the instructions are limited to the U bit and the op encoding, whereas the base encoding has bits set spread throughout (I separated the bits based on the boxes/pictures in the Arm ARM). I could make a constant for `UBitOffset`.